### PR TITLE
[Merged by Bors] - feat(Order/OmegaCompletePartialOrder): Show that Scott Continuity implies OmegaCompletePartialOrder.Continuous'

### PR DESCRIPTION
**Pull request by mans0954**
In https://github.com/leanprover-community/mathlib/pull/18517 we introduced the notion of a Scott Continuous function between preorders. This PR shows that a Scott Continuous function between OmegaCompletePartialOrders is necessarily `OmegaCompletePartialOrder.Continuous'`.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Order/OmegaCompletePartialOrder.lean [hunk 1]
```diff
@@ -232,6 +235,25 @@ theorem ωSup_le_iff (c : Chain α) (x : α) : ωSup c ≤ x ↔ ∀ i, c i ≤
   exact ωSup_le _ _ ‹_›
 #align omega_complete_partial_order.ωSup_le_iff OmegaCompletePartialOrder.ωSup_le_iff
 
+lemma ωSup_IsLUB {c : Chain α} : IsLUB (Set.range c) (ωSup c) := by
```
Comment (jcommelin) [inline Mathlib/Order/OmegaCompletePartialOrder.lean:238]:
```suggestion
lemma IsLUB_range_ωSup (c : Chain α) : IsLUB (Set.range c) (ωSup c) := by
```

File: Mathlib/Order/OmegaCompletePartialOrder.lean [hunk 2]
```diff
@@ -232,6 +235,25 @@ theorem ωSup_le_iff (c : Chain α) (x : α) : ωSup c ≤ x ↔ ∀ i, c i ≤
...
+  constructor
+  · simp only [upperBounds, Set.mem_range, forall_exists_index, forall_apply_eq_imp_iff',
+      Set.mem_setOf_eq]
+    exact fun a ↦ le_ωSup c a
+  · simp only [lowerBounds, upperBounds, Set.mem_range, forall_exists_index,
+      forall_apply_eq_imp_iff', Set.mem_setOf_eq]
+    exact fun ⦃a⦄ a_1 ↦ ωSup_le c a a_1
+
+lemma IsLUB_ωSup {c : Chain α} {a : α} (h: IsLUB (Set.range c) a) : a = ωSup c := by
```
Comment (jcommelin) [inline Mathlib/Order/OmegaCompletePartialOrder.lean:247]:
```suggestion
lemma ωSup_eq_of_IsLUB {c : Chain α} {a : α} (h : IsLUB (Set.range c) a) : a = ωSup c := by
```

File: Mathlib/Order/OmegaCompletePartialOrder.lean [hunk 3]
```diff
@@ -263,6 +285,21 @@ def Continuous' (f : α → β) : Prop :=
...
+lemma IsLUB_of_ScottContinuous {c : Chain α} {f : α → β} (hf : ScottContinuous f) :
+    IsLUB (Set.range (Chain.map c ⟨f, (ScottContinuous.monotone hf)⟩)) (f (ωSup c)) := by
+  simp only [map_coe, OrderHom.coe_mk]
+  rw [(Set.range_comp f ↑c)]
+  apply hf
+  exact Set.range_nonempty ↑c
+  exact IsChain.directedOn (isChain_range c)
+  exact ωSup_IsLUB
+
+lemma Continuous'_of_ScottContinuous {f : α → β} (hf : ScottContinuous f) : Continuous' f := by
```
Comment (jcommelin) [inline Mathlib/Order/OmegaCompletePartialOrder.lean:297]:
```suggestion
lemma ScottContinuous.continuous' {f : α → β} (hf : ScottContinuous f) : Continuous' f := by
```

**Review by jcommelin — COMMENTED** [review]
Some stylistic remarks.

Note that on L303 the name is also wrong. It should simply be `.monotone` instead of `.to_monotone`. But I don't claim this should be fixed in this PR.

**Comment by mans0954** [comment]
```
No space left on device : '/home/lean/actions-runner/_work/_actions/dcarbone/install-jq-action/v1.0.1'
```
Not sure there's anything I can do about this @jcommelin ?

**Comment by mans0954** [comment]
> ```
> No space left on device : '/home/lean/actions-runner/_work/_actions/dcarbone/install-jq-action/v1.0.1'
> ```
Possibly fixed by #7166?

File: Mathlib/Order/OmegaCompletePartialOrder.lean [hunk 4]
```diff
@@ -108,10 +108,10 @@ variable (c c' : Chain α)
 variable (f : α →o β)
 variable (g : β →o γ)
 
-instance : LE (Chain α) where le x y := ∀ i, ∃ j, x i ≤ y j
-
 lemma isChain_range : IsChain (· ≤ ·) (Set.range c) := Monotone.isChain_range (OrderHomClass.mono c)
 
+instance : LE (Chain α) where le x y := ∀ i, ∃ j, x i ≤ y j
```
Comment (jcommelin) [inline Mathlib/Order/OmegaCompletePartialOrder.lean:113]: Why are you moving this instance?

File: Mathlib/Order/OmegaCompletePartialOrder.lean [hunk 5]
```diff
@@ -268,6 +287,21 @@ def Continuous' (f : α → β) : Prop :=
...
 #align omega_complete_partial_order.continuous' OmegaCompletePartialOrder.Continuous'
 
+lemma IsLUB_of_ScottContinuous {c : Chain α} {f : α → β} (hf : ScottContinuous f) :
+    IsLUB (Set.range (Chain.map c ⟨f, (ScottContinuous.monotone hf)⟩)) (f (ωSup c)) := by
+  simp only [map_coe, OrderHom.coe_mk]
+  rw [(Set.range_comp f ↑c)]
+  apply hf
+  exact Set.range_nonempty ↑c
+  exact IsChain.directedOn (isChain_range c)
+  exact IsLUB_range_ωSup c
```
Comment (jcommelin) [inline Mathlib/Order/OmegaCompletePartialOrder.lean:297]: Could you please format this with subgoals `·` or move them as inline arguments in the place where the corresponding goal is created?

File: Mathlib/Order/OmegaCompletePartialOrder.lean (hunk 4 above)
Comment (mans0954) [reply Mathlib/Order/OmegaCompletePartialOrder.lean]: Sorry - that crept in during a merge conflict.

File: Mathlib/Order/OmegaCompletePartialOrder.lean (hunk 5 above)
Comment (mans0954) [reply Mathlib/Order/OmegaCompletePartialOrder.lean]: Moved inline.

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**